File: ts/src/conditional-groups.ts
/* Copyright (c) 2026 Richard Rodger and contributors, MIT License */

// Conditional-group folding: a translation-unit-level post-pass that
// collapses contiguous runs of `#if`/`#ifdef`/`#ifndef` … `#elif` …
// `#else` … `#endif` directives into a single conditional_group node.
// Best-effort: an unmatched `#endif` or an unterminated `#if` leaves
// the surrounding children unchanged so the rest of the tree stays
// intact.
//
// The walker is structural — it only inspects already-parsed
// conditional_directive nodes embedded as the first child of an
// external_declaration — so it has zero dependency on the token
// stream or the rest of the grammar.

// Output shape:
//   conditional_group
//     branches: [
//       { kind: 'conditional_branch', branchKind: 'if'|'ifdef'|'ifndef'|
//                                                   'elif'|'elifdef'|
//                                                   'elifndef'|'else',
//         children: [<the directive node>, <body external_declarations>] }
//       ...
//     ]
//     endif: <the conditional_directive node, kept verbatim>

interface AnyNode {
  kind?: string
  span?: any
  children?: any[]
  branches?: any[]
  endif?: any
  branchKind?: string
  directive?: any
  body?: any[]
  [extra: string]: any
}

function makeNode(kind: string, span?: any): AnyNode {
  return {
    kind,
    span: span || { start: 0, end: 0, line: 1, col: 1 },
    children: [],
    trivia: { leading: [], trailing: [] },
  } as AnyNode
}

export function structureConditionalGroups(parent: AnyNode): void {
  if (!Array.isArray(parent.children)) return
  const ch = parent.children
  const out: any[] = []
  let i = 0
  while (i < ch.length) {
    const c = ch[i] as any
    const dir = leadingConditionalDirective(c)
    if (dir && /^(if|ifdef|ifndef)$/.test(dir.directive)) {
      const consumed = tryBuildConditionalGroup(ch, i)
      if (consumed) {
        out.push(consumed.node)
        i = consumed.next
        continue
      }
    }
    out.push(c)
    i++
  }
  parent.children = out
  // Recurse into preserved children (e.g. function bodies).
  for (const c of out) {
    if (c && c.kind && Array.isArray(c.children)) {
      structureConditionalGroups(c)
    }
  }
}

// Return the conditional_directive node embedded as the first child of
// an external_declaration, if any.
function leadingConditionalDirective(node: any): any | null {
  if (!node || node.kind !== 'external_declaration') return null
  const first = (node.children || []).find(
    (c: any) => c && c.kind === 'conditional_directive',
  )
  return first || null
}

// Attempt to build a conditional_group starting at index `from`. Returns
// the new node and the index after the closing #endif, or null if no
// matching #endif was found at the same nesting level.
function tryBuildConditionalGroup(
  children: any[], from: number,
): { node: AnyNode; next: number } | null {
  // First pass: scan ahead with a depth counter to find the matching
  // #endif. If none, bail.
  let depth = 0
  let endIdx = -1
  for (let i = from; i < children.length; i++) {
    const dir = leadingConditionalDirective(children[i])
    if (!dir) continue
    if (/^(if|ifdef|ifndef)$/.test(dir.directive)) depth++
    else if (dir.directive === 'endif') {
      depth--
      if (depth === 0) { endIdx = i; break }
    }
  }
  if (endIdx < 0) return null

  // Second pass: split [from … endIdx-1] into branches at top-level
  // #elif/#else.
  const startCh = children[from] as AnyNode
  const groupNode = makeNode('conditional_group', startCh.span)
  groupNode.branches = [] as any[]
  let branchStart = from
  let innerDepth = 0
  for (let i = from + 1; i < endIdx; i++) {
    const dir = leadingConditionalDirective(children[i])
    if (!dir) continue
    if (/^(if|ifdef|ifndef)$/.test(dir.directive)) innerDepth++
    else if (dir.directive === 'endif') innerDepth--
    else if (innerDepth === 0 &&
             /^(elif|elifdef|elifndef|else)$/.test(dir.directive)) {
      pushBranch(groupNode, children, branchStart, i)
      branchStart = i
    }
  }
  pushBranch(groupNode, children, branchStart, endIdx)

  // Endif as a separate field (the directive itself, for full fidelity).
  groupNode.endif = children[endIdx]
  // Also append it to children so a depth-first walk still emits the
  // raw tokens in order.
  groupNode.children!.push(...groupNode.branches!)
  groupNode.children!.push(children[endIdx])

  return { node: groupNode, next: endIdx + 1 }
}

function pushBranch(group: AnyNode, children: any[], from: number, to: number): void {
  const head = children[from]
  const dir = leadingConditionalDirective(head)
  const branch = makeNode('conditional_branch', head.span)
  branch.branchKind = dir ? dir.directive : 'unknown'
  // The directive itself is preserved on a side field; the branch's
  // `children` list holds just the body items so consumers can iterate
  // them without filtering.
  branch.directive = head
  // Recurse into body so nested #if … #endif inside a branch also get
  // grouped.
  const inner = makeNode('__branch_body__', head.span)
  for (let k = from + 1; k < to; k++) inner.children!.push(children[k])
  structureConditionalGroups(inner)
  // Final children: the directive followed by the (possibly grouped)
  // body, so a depth-first walk still yields original tokens.
  branch.children = [head, ...inner.children!]
  // Body-only view for consumers.
  branch.body = inner.children!
  group.branches!.push(branch)
}

File: go/conditional_groups.go
/* Copyright (c) 2026 Richard Rodger and contributors, MIT License */

package tabnasc

// Conditional-group folding: a translation-unit-level post-pass that collapses
// contiguous runs of #if/#ifdef/#ifndef … #elif … #else … #endif directives
// into a single conditional_group node. Port of ../ts/src/conditional-groups.ts.
//
// Best-effort: an unmatched #endif or unterminated #if leaves the surrounding
// children unchanged. The walker is structural — it only inspects already-
// parsed conditional_directive nodes embedded as the first child of an
// external_declaration — so it has zero dependency on the token stream.

func isIfOpen(d string) bool {
	return d == "if" || d == "ifdef" || d == "ifndef"
}

func isElseLike(d string) bool {
	return d == "elif" || d == "elifdef" || d == "elifndef" || d == "else"
}

// structureConditionalGroups folds conditional directives in parent.children
// in place, then recurses into preserved children.
func structureConditionalGroups(parent CNode) {
	chRaw, ok := parent["children"].([]any)
	if !ok {
		return
	}
	out := make([]any, 0, len(chRaw))
	i := 0
	for i < len(chRaw) {
		c := chRaw[i]
		dir := leadingConditionalDirective(c)
		if dir != nil {
			if d, _ := dir["directive"].(string); isIfOpen(d) {
				if node, next, built := tryBuildConditionalGroup(chRaw, i); built {
					out = append(out, node)
					i = next
					continue
				}
			}
		}
		out = append(out, c)
		i++
	}
	parent["children"] = out
	// Recurse into preserved children (e.g. function bodies).
	for _, c := range out {
		if cm, ok := c.(CNode); ok {
			if _, hasKind := cm["kind"]; hasKind {
				if _, hasCh := cm["children"].([]any); hasCh {
					structureConditionalGroups(cm)
				}
			}
		}
	}
}

// leadingConditionalDirective returns the conditional_directive node embedded
// as the first child of an external_declaration, or nil.
func leadingConditionalDirective(node any) CNode {
	nm, ok := node.(CNode)
	if !ok || nm["kind"] != "external_declaration" {
		return nil
	}
	children, ok := nm["children"].([]any)
	if !ok {
		return nil
	}
	for _, c := range children {
		if cm, ok := c.(CNode); ok && cm["kind"] == "conditional_directive" {
			return cm
		}
	}
	return nil
}

// tryBuildConditionalGroup attempts to build a conditional_group starting at
// index `from`. Returns the new node, the index after the closing #endif, and
// true; or false if no matching #endif at the same nesting level.
func tryBuildConditionalGroup(children []any, from int) (CNode, int, bool) {
	// First pass: find the matching #endif by depth.
	depth := 0
	endIdx := -1
	for i := from; i < len(children); i++ {
		dir := leadingConditionalDirective(children[i])
		if dir == nil {
			continue
		}
		d, _ := dir["directive"].(string)
		if isIfOpen(d) {
			depth++
		} else if d == "endif" {
			depth--
			if depth == 0 {
				endIdx = i
				break
			}
		}
	}
	if endIdx < 0 {
		return nil, 0, false
	}

	startCh, _ := children[from].(CNode)
	groupNode := makeNode("conditional_group", spanOfNode(startCh))
	branches := []any{}

	// Second pass: split [from … endIdx-1] into branches at top-level
	// #elif/#else.
	branchStart := from
	innerDepth := 0
	for i := from + 1; i < endIdx; i++ {
		dir := leadingConditionalDirective(children[i])
		if dir == nil {
			continue
		}
		d, _ := dir["directive"].(string)
		if isIfOpen(d) {
			innerDepth++
		} else if d == "endif" {
			innerDepth--
		} else if innerDepth == 0 && isElseLike(d) {
			branches = append(branches, buildBranch(children, branchStart, i))
			branchStart = i
		}
	}
	branches = append(branches, buildBranch(children, branchStart, endIdx))

	groupNode["branches"] = branches
	groupNode["endif"] = children[endIdx]
	// Append branches + endif to children so a depth-first walk still emits
	// the raw tokens in order.
	kids := groupNode["children"].([]any)
	kids = append(kids, branches...)
	kids = append(kids, children[endIdx])
	groupNode["children"] = kids

	return groupNode, endIdx + 1, true
}

// buildBranch constructs a conditional_branch node for children[from:to].
func buildBranch(children []any, from, to int) CNode {
	head, _ := children[from].(CNode)
	dir := leadingConditionalDirective(children[from])
	branch := makeNode("conditional_branch", spanOfNode(head))
	if dir != nil {
		branch["branchKind"] = dir["directive"]
	} else {
		branch["branchKind"] = "unknown"
	}
	branch["directive"] = head
	// Recurse into body so nested #if … #endif inside a branch also group.
	inner := makeNode("__branch_body__", spanOfNode(head))
	innerKids := []any{}
	for k := from + 1; k < to; k++ {
		innerKids = append(innerKids, children[k])
	}
	inner["children"] = innerKids
	structureConditionalGroups(inner)
	body := inner["children"].([]any)
	// Final children: directive followed by the (possibly grouped) body.
	final := make([]any, 0, 1+len(body))
	final = append(final, head)
	final = append(final, body...)
	branch["children"] = final
	branch["body"] = body
	return branch
}

// spanOfNode returns n["span"] as a map, or nil.
func spanOfNode(n CNode) map[string]any {
	if n == nil {
		return nil
	}
	if s, ok := n["span"].(map[string]any); ok {
		return s
	}
	return nil
}
